boost::spirit::x3 problem with any ast class representing a std::string

I have two, what I presume to be legitimate, ways of defining a class in my ast that represents a string, illustrated here:
```
  struct white : std::string {};
```
or
```
  struct white {std::string text;};
```
which I use in several places for things like literals, identifiers, comments, or even white space, which I want to capture in my ast.

In some places I have to use the first form, and in other places I have to use the second form, otherwise I get compile errors with horrendous error messages:
```
  ...
  no matching function for call to ??minimal::ast::rules::rules(__gnu_cxx::__normal_iterator<minimal::ast::gap_item*,
    std::vector<minimal::ast::gap_item> >&,
    __gnu_cxx::__normal_iterator<minimal::ast::gap_item*, std::vector<minimal::ast::gap_item> >&)??
  ...
```
embedded in several pages of error output with no clue regarding the line of source code which might be at fault.

How should I interpret this type of error message?
Why do I have to use a different form/style of class in different places?

This minimal example illustrates the problem (when the #defines are [un]commented)
[rgw29_minimal.cpp.txt](https://github.com/boostorg/spirit/files/6518713/rgw29_minimal.cpp.txt)

@Kojoley I naively assumed that the MCVE you provided above was a working/compilable example of correct coding.

However, when I attempt to compile this MCVE I get:
```
  src/rgw29_nikita_mcve.cpp:58:41:   required from here
  /usr/include/boost/spirit/home/x3/support/ast/variant.hpp:151:39: error: no matching function for call to ??boost::variant<minimal::ast::gap, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::variant(minimal::ast::gap_item&)??
    151 |             : var(std::forward<T>(rhs)) {}
        |                                       ^
```
Perhaps this is what you intended.  But since this MCVE does not contain any single element structs (unsupported single element tuple) and each derived struct does have constructors and assignment methods, I cannot see what is wrong.

Please could you provide a novice x3 user level explanation for why this code fails to compile and what would be an appropriate fix.

I have another example, which I believe to be suffering from the same problem.
_But I am sure you will direct me to place this in a separate issue, if not._

```
#define BOOST_SPIRIT_X3_DEBUG	// Provide some meaningful reassuring output.
#define WITH_TERMINATOR		// 
#define WITH_GAP_ITEM		//

#include <string>
#include <boost/spirit/home/x3.hpp>
#include <boost/spirit/home/x3/support/ast/variant.hpp>
#include <boost/fusion/include/adapt_struct.hpp>

namespace x3		= boost::spirit::x3;

namespace minimal { namespace ast
{
  struct braced						{char open; std::string	text; char close;};
  struct gap_item	: x3::variant<braced,char>	{using base_type::operator=;};
  struct gap 		: std::vector<gap_item>		{};

  struct start_item	: x3::variant< std::string
  #ifdef WITH_GAP_ITEM
  , gap_item
  #else
  , gap
  #endif
  #ifdef WITH_TERMINATOR
  , char
  #endif
						>	{using base_type::base_type; using base_type::operator=;};
  struct start_rule	: std::vector<start_item>	{using std::vector<start_item>::vector;};
}}
BOOST_FUSION_ADAPT_STRUCT(minimal::ast::braced, open, text, close)

namespace minimal { namespace parser
{
  using x3::char_;
  using x3::space;
  
  x3::rule<struct braced,	ast::braced>	 const	braced		= "braced";
  x3::rule<struct gap,		ast::gap>	 const	gap		= "gap";
  x3::rule<struct start_rule,	ast::start_rule> const	start_rule	= "start_rule";
  
  auto const braced_def		= char_('{') >> *(char_ -'}') >> char_('}');	// { ... }
  auto const gap_def		= +((space) | braced);				// spaces etc

  auto const start_rule_def	= +(gap | +(char_ -'{' -space -';'))		// gap or string
    #ifdef WITH_TERMINATOR
    > char_(';')
    #endif
    ;
  
  BOOST_SPIRIT_DEFINE(gap, braced, start_rule);
}}

int
main()
{
  char const*			iter	= "? {;};", * const end	= iter + std::strlen(iter);
  minimal::ast::start_rule	ast;
  return !parse(iter, end, minimal::parser::start_rule, ast) || iter!=end;
}
```
The starting point should be with these two defines commented:
```
//#define WITH_TERMINATOR		// 
//#define WITH_GAP_ITEM		//
```
Which gives a working parser which compiles, but for the input supplied, exits with a non-zero status because the end of statement ';' is missing from the grammar.

So I now add that terminator to the grammar
```
> char_(';')
```
and of course, I have to change the contents of the variant by adding ```char```.  This is done by uncommenting
```
#define WITH_TERMINATOR		// 
```
but this fails to compile, with the guts of the error message being:
```
src/rgw29_terminator.cpp:63:60:   required from here
/usr/include/boost/spirit/home/x3/support/ast/variant.hpp:151:39: error: no matching function for call to ??boost::variant<std::__cxx11::basic_string<char>, minimal::ast::gap, char>::variant(minimal::ast::gap_item&)??
  151 |             : var(std::forward<T>(rhs)) {}
      |                                       ^
```
This was a total surprise to me.  Why did adding something extra to the end of my start_rule break things so dramatically?

But I am becoming used to these error messages, and on a hunch, changed the collective ```gap``` (a ```vector<gap_item>```) to ```gap item```, acheived by uncommenting:
```
#define WITH_GAP_ITEM		//
```

Can you possibly provide an explanation for this X3 behaviour?

There's a hint of the collapsing attributes, mentioned above, with the workaround being similar: drop one level of vector nesting.
However this is *not* a viable workaround for any real world parser.

There is, I believe, a genuine problem here.  Let me illustrate with another example.

My starting point is
```
#include <string>
#include <boost/spirit/home/x3.hpp>
#include <boost/spirit/home/x3/support/ast/variant.hpp>
#include <boost/fusion/include/adapt_struct.hpp>

namespace x3		= boost::spirit::x3;

namespace minimal { namespace ast
{
  struct braced						{char open; std::string	text; char close;};
  struct gap_item	: x3::variant<char,braced>	{using base_type::operator=;};
  struct gap 		: std::vector<gap_item>		{};
  struct start_item	: x3::variant<gap,std::string>	{using base_type::operator=;};
  struct start	: std::vector<start_item>		{using std::vector<start_item>::vector;};
}}
BOOST_FUSION_ADAPT_STRUCT(minimal::ast::braced, open, text, close)

namespace minimal { namespace parser
{
  using x3::char_;
  using x3::space;
  using x3::raw;
  
  x3::rule<struct white,	ast::white>	const	white		= "white";
  x3::rule<struct braced,	ast::braced>	const	braced		= "braced";
  x3::rule<struct gap,		ast::gap>	const	gap		= "gap";
  x3::rule<struct start,	ast::start>	const	start		= "start";
  
  auto const white_def		= raw[+space];
  auto const braced_def		= char_('{') >> *(char_ -'}') >> char_('}');	// { ... }
  auto const gap_def		= +(space | braced);				// spaces etc
  auto const start_def		= +(gap | +(char_ -'{' -space));		// gap=container or string
  
  BOOST_SPIRIT_DEFINE(white, gap, braced, start);
}}

int
main()
{
  char const*			iter	= "? {;};", * const end	= iter + std::strlen(iter);
  minimal::ast::start	ast;
  return !parse(iter, end, minimal::parser::start, ast) || iter!=end;
}
```
which compiles and runs without issue.

Next I add ```white``` for the collective white space, rather than just a single ```space``` character.
```
namespace minimal { namespace ast
{
  struct white		: std::string			{using std::string::string;};
  struct braced						{char open; std::string	text; char close;};
  struct gap_item	: x3::variant<white,braced>	{using base_type::operator=;};
  struct gap 		: std::vector<gap_item>		{};
  struct start_item	: x3::variant<gap,std::string>	{using base_type::operator=;};
  struct start	: std::vector<start_item>		{using std::vector<start_item>::vector;};
}}
BOOST_FUSION_ADAPT_STRUCT(minimal::ast::braced, open, text, close)

namespace minimal { namespace parser
{
  using x3::char_;
  using x3::space;
  using x3::raw;
  
  x3::rule<struct white,	ast::white>	const	white		= "white";
  x3::rule<struct braced,	ast::braced>	const	braced		= "braced";
  x3::rule<struct gap,		ast::gap>	const	gap		= "gap";
  x3::rule<struct start,	ast::start>	const	start		= "start";
  
  auto const white_def		= raw[+space];
  auto const braced_def		= char_('{') >> *(char_ -'}') >> char_('}');	// { ... }
  auto const gap_def		= +(white | braced);				// spaces etc
  auto const start_def		= +(gap | +(char_ -'{' -space));		// gap=container or string
  
  BOOST_SPIRIT_DEFINE(white, gap, braced, start);
}}
```
However, this fails to compile giving


```
src/rgw29_gap_item.cpp:64:55:   required from here
/usr/include/c++/10/bits/stl_uninitialized.h:137:72: error: static assertion failed: result type must be
constructible from value type of input range
  137 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~
src/rgw29_gap_item.cpp:64:55:   required from here
/usr/include/boost/spirit/home/x3/support/ast/variant.hpp:184:17: error: no match for ??operator=??
(operand types are ??boost::spirit::x3::variant<minimal::ast::gap, std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> > >::variant_type?? {aka ??boost::variant<minimal::ast::gap,
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >??} and ??const char??)
  184 |             var = std::forward<T>(rhs);
      |             ~~~~^~~~~~~~~~~~~~~~~~~~~~
```
essentially telling me to add a ```char``` to my start_item variant.  And when I do that I get
```
src/rgw29_gap_item.cpp:64:55:   required from here
/usr/include/c++/10/bits/stl_uninitialized.h:137:72: error: static assertion failed: result type must be
constructible from value type of input range
  137 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~

src/rgw29_gap_item.cpp:64:55:   required from here
/usr/include/boost/spirit/home/x3/support/ast/variant.hpp:184:17: error: no match for ??operator=??
(operand types are ??boost::spirit::x3::variant<minimal::ast::gap, std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> >, char>::variant_type?? {aka ??boost::variant<minimal::ast::gap,
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>??} and
??minimal::ast::gap_item??)
  184 |             var = std::forward<T>(rhs);
      |             ~~~~^~~~~~~~~~~~~~~~~~~~~~
```
essentially telling me to add a ```gap_item``` to my start_item variant.  And when I do that I get it does compile parse ok.

But this is utter madness.  Why do I have to add ```char``` and ```gap_item``` to my variant, when that already contains ```gap``` and ```gap``` is a container which contains one or more ```char``` or ```gap_item```.  And to be clear, although this does make this minimal code compile and parse, in my real world code the ```gap``` in the ast has been bypassed and the code therein is not executed (for instance, the gap is not printed).  This is unnacceptable.

This also breaks the notion you @Kojoley suggested that
> a rule is a barrier for any trickery.

What is going wrong with X3, or what am I doing wrong?

--- a/include/boost/spirit/home/qi/detail/alternative_function.hpp
+++ b/include/boost/spirit/home/qi/detail/alternative_function.hpp
@@ -20,22 +20,22 @@
 
 namespace boost { namespace spirit { namespace qi { namespace detail
 {
-    template <typename Variant, typename Expected>
+    template <typename Variant, typename T>
     struct find_substitute
     {
-        // Get the type from the variant that can be a substitute for Expected.
-        // If none is found, just return Expected
+        // Get the type from the Variant that can be a substitute for T.
+        // If none is found, just return T
 
         typedef Variant variant_type;
         typedef typename variant_type::types types;
         typedef typename mpl::end<types>::type end;
 
-        typedef typename mpl::find<types, Expected>::type iter_1;
+        typedef typename mpl::find<types, T>::type iter_1;
 
         typedef typename
             mpl::eval_if<
                 is_same<iter_1, end>,
-                mpl::find_if<types, traits::is_substitute<mpl::_1, Expected> >,
+                mpl::find_if<types, traits::is_substitute<T, mpl::_1> >,
                 mpl::identity<iter_1>
             >::type
         iter;
@@ -43,7 +43,7 @@ namespace boost { namespace spirit { namespace qi { namespace detail
         typedef typename
             mpl::eval_if<
                 is_same<iter, end>,
-                mpl::identity<Expected>,
+                mpl::identity<T>,
                 mpl::deref<iter>
             >::type
         type;

--- a/include/boost/spirit/home/x3/support/traits/variant_find_substitute.hpp
+++ b/include/boost/spirit/home/x3/support/traits/variant_find_substitute.hpp
@@ -13,22 +13,22 @@
 
 namespace boost { namespace spirit { namespace x3 { namespace traits
 {
-    template <typename Variant, typename Attribute>
+    template <typename Variant, typename T>
     struct variant_find_substitute
     {
-        // Get the type from the variant that can be a substitute for Attribute.
-        // If none is found, just return Attribute
+        // Get the type from the Variant that can be a substitute for T.
+        // If none is found, just return T
 
         typedef Variant variant_type;
         typedef typename variant_type::types types;
         typedef typename mpl::end<types>::type end;
 
-        typedef typename mpl::find<types, Attribute>::type iter_1;
+        typedef typename mpl::find<types, T>::type iter_1;
 
         typedef typename
             mpl::eval_if<
                 is_same<iter_1, end>,
-                mpl::find_if<types, traits::is_substitute<mpl::_1, Attribute> >,
+                mpl::find_if<types, traits::is_substitute<T, mpl::_1> >,
                 mpl::identity<iter_1>
             >::type
         iter;
@@ -36,7 +36,7 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
         typedef typename
             mpl::eval_if<
                 is_same<iter, end>,
-                mpl::identity<Attribute>,
+                mpl::identity<T>,
                 mpl::deref<iter>
             >::type
         type;

--- a/include/boost/spirit/home/x3/support/traits/variant_has_substitute.hpp
+++ b/include/boost/spirit/home/x3/support/traits/variant_has_substitute.hpp
@@ -13,22 +13,22 @@
 
 namespace boost { namespace spirit { namespace x3 { namespace traits
 {
-    template <typename Variant, typename Attribute>
+    template <typename Variant, typename T>
     struct variant_has_substitute_impl
     {
-        // Find a type from the variant that can be a substitute for Attribute.
+        // Find a type from the Variant that can be a substitute for T.
         // return true_ if one is found, else false_
 
         typedef Variant variant_type;
         typedef typename variant_type::types types;
         typedef typename mpl::end<types>::type end;
 
-        typedef typename mpl::find<types, Attribute>::type iter_1;
+        typedef typename mpl::find<types, T>::type iter_1;
 
         typedef typename
             mpl::eval_if<
                 is_same<iter_1, end>,
-                mpl::find_if<types, traits::is_substitute<mpl::_1, Attribute>>,
+                mpl::find_if<types, traits::is_substitute<T, mpl::_1>>,
                 mpl::identity<iter_1>
             >::type
         iter;
@@ -36,15 +36,15 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
         typedef mpl::not_<is_same<iter, end>> type;
     };
 
-    template <typename Variant, typename Attribute>
+    template <typename Variant, typename T>
     struct variant_has_substitute
-        : variant_has_substitute_impl<Variant, Attribute>::type {};
+        : variant_has_substitute_impl<Variant, T>::type {};
 
-    template <typename Attribute>
-    struct variant_has_substitute<unused_type, Attribute> : mpl::true_ {};
+    template <typename T>
+    struct variant_has_substitute<unused_type, T> : mpl::true_ {};
 
-    template <typename Attribute>
-    struct variant_has_substitute<unused_type const, Attribute> : mpl::true_ {};
+    template <typename T>
+    struct variant_has_substitute<unused_type const, T> : mpl::true_ {};
 
 }}}}
 

--- a/test/x3/alternative.cpp
+++ b/test/x3/alternative.cpp
@@ -274,5 +274,27 @@ main()
         BOOST_TEST_EQ(s, "abcbbcd");
     }
 
+    { // conversion between alternatives
+        struct X {};
+        struct Y {};
+        struct Z {};
+        boost::variant<X, Y, Z> v;
+        boost::variant<Y, X> x{X{}};
+        v = x; // boost::variant supports that convertion
+        auto const p = 'x' >> attr(x) | 'z' >> attr(Z{});
+        BOOST_TEST(test_attr("z", p, v));
+        BOOST_TEST(boost::get<Z>(&v) != nullptr);
+        BOOST_TEST(test_attr("x", p, v));
+        BOOST_TEST(boost::get<X>(&v) != nullptr);
+    }
+
+    { // regression test for #679
+        using Qaz = std::vector<boost::variant<int>>;
+        using Foo = std::vector<boost::variant<Qaz, int>>;
+        using Bar = std::vector<boost::variant<Foo, int>>;
+        Bar x;
+        BOOST_TEST(test_attr("abaabb", +('a' >> attr(Foo{}) | 'b' >> attr(int{})), x));
+    }
+
     return boost::report_errors();
 }

